File: DGAdLaunchView ObjC/DGAdLaunchView ObjC/AdLaunchView.m
//
//  AdLaunchView.m
//  DGAdLaunchView ObjC
//
//  Created by 段昊宇 on 16/5/26.
//  Copyright © 2016年 Desgard_Duan. All rights reserved.
//

#import "AdLaunchView.h"
#import "DACircularProgressView.h"
#import "UIImageView+WebCache.h"
#import "Masonry.h"
#import "MZTimerLabel.h"

AdLaunchType state = AdLaunchProgressType;

@interface AdLaunchView() 

@property(nonatomic, strong) UIView *adBackground;
@property(nonatomic, strong) NSString *imageURL;
@property(nonatomic, strong) UIImageView *adImageView;
@property(nonatomic, strong) DACircularProgressView *progressView;
@property(nonatomic, strong) UIButton *progressButtonView;


@end

@implementation AdLaunchView

#pragma mark - override
- (instancetype) initWithFrame:(CGRect)frame {
    return [self initWithFrame: frame type: AdLaunchProgressType];
}

- (instancetype) initWithFrame: (CGRect)frame type: (AdLaunchType) type {
    state = type;
    self = [super initWithFrame: frame];
    [self addSubview: self.adBackground];
    [self displayCachedAd];
    [self requestBanner];
    [self showProgressView];
    
    dispatch_time_t show = dispatch_time(DISPATCH_TIME_NOW, 4 * NSEC_PER_SEC);
    dispatch_after(show, dispatch_get_main_queue(), ^(void){
        [self removeFromSuperview];
    });
    
    return self;
}

- (void) removeFromSuperview {
    [UIView animateWithDuration:0.8
                     animations:^{
        self.alpha = 0;
    }
                     completion:^(BOOL finished) {
        [super removeFromSuperview];
    }];
}

#pragma mark - 获取本地图片缓存，如果没有广告结束
- (void) displayCachedAd {
    SDWebImageManager *manage = [[SDWebImageManager alloc] init];
    NSURL *url = [[NSURL alloc] initWithString: self.imageURL];
    if ([manage cacheKeyForURL: url] == NO) {
        self.hidden = YES;
    } else {
        [self showImage];
    }
}

#pragma mark - 展示图片
- (void) showImage {
    self.adImageView = [[UIImageView alloc] initWithFrame: CGRectMake(0, 0, [UIScreen mainScreen].bounds.size.width, [UIScreen mainScreen].bounds.size.height - 128)];
    [self.adImageView sd_setImageWithURL: [[NSURL alloc] initWithString: self.imageURL]];
    [self.adImageView setUserInteractionEnabled: YES];
    UITapGestureRecognizer *singleTap = [[UITapGestureRecognizer alloc] initWithTarget: self action: @selector(singleTap)];
    [self.adImageView addGestureRecognizer: singleTap];
    [self addSubview: _adImageView];
}

#pragma mark - 图片点击事件
- (void) singleTap {
    [self._delegate adLaunch: self];
    [self toHidenState];
}

#pragma mark - 下载图片
- (void) requestBanner {
    [[SDWebImageManager sharedManager] downloadImageWithURL: [[NSURL alloc] initWithString: self.imageURL]
                                                    options: SDWebImageAvoidAutoSetImage
                                                   progress: nil
                                                  completed: ^(UIImage *image, NSError *error, SDImageCacheType cacheType, BOOL finished, NSURL *imageURL) {
                                                      NSLog(@"图片下载成功");
                                                  }];
}

#pragma mark - 展示跳过按钮
- (void) showProgressView {
    switch (state) {
        case AdLaunchProgressType:
            self.progressButtonView = [[UIButton alloc] initWithFrame: CGRectMake([UIScreen mainScreen].bounds.size.width - 60, 20, 40, 40)];
            [self.progressButtonView setTitle: @"跳" forState: UIControlStateNormal];
            self.progressButtonView.titleLabel.textAlignment = NSTextAlignmentCenter;
            self.progressButtonView.backgroundColor = [UIColor clearColor];
            [self.progressButtonView addTarget: self
                                        action: @selector(toHidenState)
                              forControlEvents: UIControlEventTouchUpInside];
            [self addSubview: self.progressButtonView];
            
            self.progressView = [[DACircularProgressView alloc] initWithFrame: CGRectMake([UIScreen mainScreen].bounds.size.width - 60, 20, 40, 40)];
            self.progressView.userInteractionEnabled = NO;
            self.progressView.progress = 0;
            [self addSubview: self.progressView];
            [self.progressView setProgress: 1 animated: YES initialDelay: 0 withDuration:4];
            break;
            
        case AdLaunchTimerType:
            self.progressButtonView = [UIButton buttonWithType: UIButtonTypeRoundedRect];
            self.progressButtonView.layer.masksToBounds = YES;
            self.progressButtonView.layer.cornerRadius = 12;
            self.progressButtonView.backgroundColor = [UIColor blackColor];
            self.progressButtonView.alpha = 0.3;
            [self.progressButtonView addTarget: self
                                        action: @selector(toHidenState)
                              forControlEvents: UIControlEventTouchUpInside];
            
            
            MZTimerLabel *timeLabel = [[MZTimerLabel alloc] initWithTimerType: MZTimerLabelTypeTimer];
            timeLabel.textAlignment = NSTextAlignmentRight;
            timeLabel.timeFormat = @"s 跳过";
            timeLabel.textColor = [UIColor whiteColor];
            timeLabel.font = [UIFont systemFontOfSize: 14];
            timeLabel.userInteractionEnabled = NO;
            [timeLabel setCountDownTime: 4];
            
            [timeLabel start];
            [self addSubview: self.progressButtonView];
            [self addSubview: timeLabel];
            
            [self.progressButtonView mas_makeConstraints:^(MASConstraintMaker *make) {
                make.size.mas_equalTo(CGSizeMake(64, 24));
                make.top.equalTo(self).offset(24);
                make.right.equalTo(self).offset(-16);
            }];
            
            [timeLabel mas_makeConstraints:^(MASConstraintMaker *make) {
                make.center.equalTo(self.progressButtonView);
            }];
            
            break;
    }
    
    

}

#pragma mark - 消失动画
- (void) toHidenState {
    [UIView animateWithDuration: 0.4
                     animations:^{
                         self.alpha = 0;
                     }
                     completion:^(BOOL finished) {
                         self.hidden = true;
                     }];
}

#pragma mark - 懒加载
- (UIView *) adBackground {
    CGRect  scr = [UIScreen mainScreen].bounds;
    CGFloat wid = scr.size.width;
    CGFloat hei = scr.size.height;
    
    UIView *footer = [[UIView alloc] initWithFrame: CGRectMake(0, hei - 128, wid, 128)];
    footer.backgroundColor = [UIColor whiteColor];
    
    UIImageView *slogan = [[UIImageView alloc] initWithImage: [UIImage imageNamed: @"KDTKLaunchSlogan_Content"]];
    [footer addSubview: slogan];
    
    [slogan mas_makeConstraints:^(MASConstraintMaker *make) {
        make.center.equalTo(footer);
    }];
    
    UIView *view = [[UIView alloc] initWithFrame: [UIScreen mainScreen].bounds];
    view.backgroundColor = [UIColor whiteColor];
    [view addSubview: footer];
    return view;

    return nil;
}

- (NSString *) imageURL {
    NSString *str = @"http://www.bizhituku.com/file/d/dongmanbizhi/20150309/201503091827501686.jpg";
    return str;
}

@end

File: DGAdLaunchView/DGAdLaunchView/AdLaunchView.swift
//
//  AdLaunchView.swift
//  DGAdLaunchView
//
//  Created by Desgard_Duan on 16/5/23.
//  Copyright © 2016年 Desgard_Duan. All rights reserved.
//

import UIKit


@objc protocol AdLaunchViewDelegate: NSObjectProtocol {
    
}

final class AdLaunchView: UIView {
    
    weak var delegate: AdLaunchViewDelegate?
    
    // 启动广告背景
    private lazy var adBackground: UIView = {
        let wid = UIScreen.mainScreen().bounds.width
        let hei = UIScreen.mainScreen().bounds.height
        
        var footer: UIView = UIView(frame: CGRectMake(0, hei - 128, wid, 128))
        footer.backgroundColor = UIColor.whiteColor()
        
        var slogan: UIImageView = UIImageView(image: UIImage(named: "KDTKLaunchSlogan_Content"))
        footer.addSubview(slogan)
        
        slogan.snp_makeConstraints(closure: { (make) in
            make.center.equalTo(footer)
        })
        
        var view: UIView = UIView(frame: UIScreen.mainScreen().bounds)
        view.backgroundColor = UIColor.whiteColor()
        view.addSubview(footer)
        return view
    }()
    
    // 图片链接
    private var imageURL: String = "http://mg.soupingguo.com/bizhi/big/10/258/043/10258043.jpg"
    
    // 启动页广告
    private var adImageView: UIImageView?
    
    // 进度条
    private var progressView: DACircularProgressView?
    
    // 跳过广告按钮
    private var progressButtonView: UIButton?
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        addSubview(adBackground)
        
        // 广告主流程
        displayCachedAd()
        requestBanner()
        showProgressView()
        
        let delayTime = dispatch_time(DISPATCH_TIME_NOW, Int64(4 * Double(NSEC_PER_SEC)))
        dispatch_after(delayTime, dispatch_get_main_queue()) {
            self.removeFromSuperview()
        }
    }
    
    override func removeFromSuperview() {
        UIView.animateWithDuration(1, animations: {
                self.alpha = 0
            }) { (finished: Bool) in
                super.removeFromSuperview()
            }
    }
    
    required init?(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
}

// MARK: - private
private extension AdLaunchView {
    
    func displayCachedAd() {
        let manange: SDWebImageManager = SDWebImageManager()
        let url = NSURL(string: imageURL)
        if (!manange.cachedImageExistsForURL(url)) {
            self.hidden = true
        } else {
            showImage()
        }
    }
    
    func requestBanner() {
        SDWebImageDownloader.sharedDownloader().downloadImageWithURL(
            NSURL(string: imageURL),
            options: SDWebImageDownloaderOptions.UseNSURLCache,
            progress: nil,
            completed: { (image, data, error, bool) -> Void in
                if image != nil {
                    print("广告已经下载成功")
                }
        })
        
        
    }
    
    func showImage() {
        adImageView = UIImageView(frame: CGRect(x: 0, y: 0, width: UIScreen.mainScreen().bounds.width, height: UIScreen.mainScreen().bounds.height - 128))
        if let adImageView = adImageView {
            adImageView.sd_setImageWithURL(NSURL(string: imageURL))
            adImageView.userInteractionEnabled = true
            let singleTap: UITapGestureRecognizer = UITapGestureRecognizer(target: self, action: #selector(AdLaunchView.singleTapAction))
            adImageView.addGestureRecognizer(singleTap)
            
            addSubview(adImageView)
        }
    }
    
    func showProgressView() {
        progressButtonView = UIButton(frame: CGRectMake(UIScreen.mainScreen().bounds.width - 60, 20, 40, 40))
        if let progressButtonView = progressButtonView {
            progressButtonView.setTitle("跳", forState: .Normal)
            progressButtonView.titleLabel?.textAlignment = .Center
            progressButtonView.backgroundColor = UIColor.clearColor()
            progressButtonView.addTarget(self, action: #selector(toHidenState), forControlEvents: .TouchUpInside)
            addSubview(progressButtonView)
        }
        
        progressView = DACircularProgressView(frame: CGRectMake(UIScreen.mainScreen().bounds.width - 60, 20, 40, 40))
        if let progressView = progressView {
            progressView.userInteractionEnabled = false
            progressView.progress = 0
            addSubview(progressView)
            progressView.setProgress(1, animated: true, initialDelay: 0, withDuration: 4)
        }
    }
    
    @objc func singleTapAction() {
        
    }
    
    @objc func toHidenState() {
        UIView.animateWithDuration(0.3, animations: {
            self.alpha = 0
        }) { (finished: Bool) in
            self.hidden = true
        }
    }
}
